src/Spago/Command/Build.purs

```diff
@@ -137,26 +138,32 @@ run opts = do
       Cmd.exec backend.cmd (addOutputArgs moreBackendArgs) Cmd.defaultExecOptions >>= case _ of
         Left r -> do
           logDebug $ Cmd.printExecResult r
-          die [ "Failed to build with backend " <> backend.cmd ]
-        Right _ ->
+          prepareToDie [ "Failed to build with backend " <> backend.cmd ]
+          pure false
+        Right _ -> do
           logSuccess "Backend build succeeded."
-
-  let
-    pedanticPkgs = NEA.toArray selectedPackages # Array.mapMaybe \p -> do
-      let reportSrc = pedanticPackages || (fromMaybe false $ p.package.build >>= _.pedantic_packages)
-      let reportTest = pedanticPackages || (fromMaybe false $ p.package.test >>= _.pedantic_packages)
-      Alternative.guard (reportSrc || reportTest)
-      pure $ Tuple p { reportSrc, reportTest }
-  unless (Array.null pedanticPkgs || opts.depsOnly) do
-    logInfo $ "Looking for unused and undeclared transitive dependencies..."
-    eitherGraph <- Graph.runGraph globs opts.pursArgs
-    graph <- either die pure eitherGraph
-    env <- ask
-    checkResults <- map Array.fold $ for pedanticPkgs \(Tuple selected options) -> do
-      Graph.toImportErrors selected options
-        <$> runSpago (Record.union { selected, workspacePackages: selectedPackages } env) (Graph.checkImports graph)
-    unless (Array.null checkResults) do
-      die $ Graph.formatImportErrors checkResults
+          pure true
+
+  if not backendBuilt then
+    pure false
+  else do
+    let
+      pedanticPkgs = NEA.toArray selectedPackages # Array.mapMaybe \p -> do
+        let reportSrc = pedanticPackages || (fromMaybe false $ p.package.build >>= _.pedantic_packages)
+        let reportTest = pedanticPackages || (fromMaybe false $ p.package.test >>= _.pedantic_packages)
+        Alternative.guard (reportSrc || reportTest)
+        pure $ Tuple p { reportSrc, reportTest }
+    unless (Array.null pedanticPkgs || opts.depsOnly) do
+      logInfo $ "Looking for unused and undeclared transitive dependencies..."
+      eitherGraph <- Graph.runGraph globs opts.pursArgs
+      graph <- either die pure eitherGraph
+      env <- ask
+      checkResults <- map Array.fold $ for pedanticPkgs \(Tuple selected options) -> do
+        Graph.toImportErrors selected options
+          <$> runSpago (Record.union { selected, workspacePackages: selectedPackages } env) (Graph.checkImports graph)
+      unless (Array.null checkResults) do
+        die $ Graph.formatImportErrors checkResults
```

Comment:
Should this be a `prepareToDie` as well?

Reply:
Yes, this could be, I just didn't think hard enough to understand and restructure this code. I guess the types will make it clear enough